### PR TITLE
Parpack (second try)

### PR DESCRIPTION
**Pull request by davydden**
this superseeds https://github.com/dealii/dealii/pull/265. 

It shall not be merged as i believe there is a SEGV error in `scalapack 2.0.2`. The tests run fine on `head` version, though. 

In February i created a bug report in Lapack forum http://icl.cs.utk.edu/lapack-forum/viewtopic.php?f=13&t=4676&p=11434#p11434 but so far no reply. 
In that post there is a link to a small program of another guy, which fails for me at exactly the same place - within `scalapack`. That gives a bit of confidence that the code in this PR is likely ok.
I asked the same question on Arpack forum: http://forge.scilab.org/index.php/p/arpack-ng/issues/1563/

I create a PR for two reasons: 
1) so that it's known that the interface has been developed, but it does not work arguably due to third party libraries;
2) if anybody has a working `parpack` code in c++, i would be happy to compile it on my machine and see if I hit the same segmentation error. That would either confirm my experience or will be of help in debugging this PR.

### CODE REVIEW

**Comment by davydden** [comment]
**important update** i switched step-36 to parallel SLEPc and it runs fine with `KrylovSchur`. However, when I override deal.II SLEPc wrappers from a command line `-eps_type arpack` - it fails with the segmentation fault. `lldb` again leads to the same place. In other words, SLEPc wrappers of Arpack also produce the same error as I experience using this class in the PR.

**Comment by tamiko** [comment]
It might be worthwhile to fix the current issues with arpack first before introducing a new interface for parpack. E.g., it might be possible to combine both interfaces into one deal.II class (as the parpack variant quite resembles the arpack version).

So, I suggest to let this PR open for 2 weeks and discuss about this issue a bit in College Station...

@guidokanschat @tjhei @bangerth Opinions?

**Comment by bangerth** [comment]
It's not my area, so I won't have little to contribute. But I'm happy to facilitate!

**Comment by guidokanschat** [comment]
Denis, what does Parpack do that Arpack doesn't. The solvers and matrix vector products going into Arpack can be parallel. What is left is computing eigenvalues of the projected system, which we can consider small. Do you know what the additional parallelism is?

**Comment by davydden** [comment]
Guido, i am not sure that can be considered small, i think it depends. I did not look into details if implementation (if they are documented at all), but I can speculate that additional parallelism is related to operations on vectors, as effectively locally owned vectors are provided as an input. You are right that something like `H * y` and `M * y` is done outside, but when inside `(p)arpack` will compute `x * (H * y)` and `x * (M * y)`. Hence in serial one would waste memory and those operations (and alike) wouldn't scale well either. 

Obviously, i did not do any study to see at what point (in terms of system size, number of eigenpairs to be computed and number of mpi processes, presence or absence of spectral transformation) one sees advantage of `parpack`, but since parallel matrix-vector products are there in deal.ii, i would argue that it would be good to have `parpack` working.

**Comment by guidokanschat** [comment]
Thanks, Denis. I see. Actually, you could indeed ask for millions of eigenvalues and need parallelism just to store all the Schur vectors.

I am wondering right now: who tells parpack how the vectors are distributed. Or does it redistribute parallel vectors?

**Comment by davydden** [comment]
one provides to parpack something like `n_local_dofs`. How exactly they add-up to a global vector (i.e. `IndexSet locally_owned_dofs`) should not matter, IMHO, for operations like scalar products and subtractions.

**Comment by davydden** [comment]
i think i nailed what causes the problem within `parpack` https://github.com/opencollab/arpack-ng/issues/18#issuecomment-123506890 . 
Why does it lead to a segmentation fault - i don't know. My `(p)arpack` (with `MPI`) ends up linked against `Scalapack` which inside its `pdlamch` calls  `BLACS`'s `dgamx2d` . Could that be that `BLACS` and `MPI` do not like each other somehow?

p.s. keep in mind that arpack is not developed since 98. Current maintainers are Debian, Octave and Scilab.

**Comment by davydden** [comment]
i just spoke with Wolfang @bangerth . One possible solution is to add a configuration test which compiles and runs 
````
! compile as :
! mpif90 -o hello -L/usr/local/opt/scalapack/lib/ -lscalapack hello.f90
program hello

double precision pdlamch
external   pdlamch

include   'mpif.h'

integer    comm
integer ierr

double precision eps23

call MPI_INIT ( ierr )
call MPI_Comm_size(MPI_COMM_WORLD, nPEs, ierr)

comm = MPI_COMM_WORLD

print *, 'Here comes a segmentation fault!'
! eps23 = pdlamch(MPI_COMM_WORLD, 'Epsilon-Machine')
eps23 = pdlamch(comm, 'E')
print *, 'eps23=',eps23

call MPI_FINALIZE ( ierr )
stop

end program hello
```
If it is good, the Parpack should work and therefore the interface here should work. 
Otherwise, parpack interface should be disabled.

@tamiko what do you think?

p.s. As far as I understand the failing function in Scalapack (`pdlamch`) also exists in lapack alone, so if users don't have scalapack installed, it actually should be fine, i suppose.

**Comment by davydden** [comment]
one could also run one of the tests in this PR and see if they run. If they do, it should be good and we can keep Parpack interface.

**Comment by davydden** [comment]
uninstalled `scalapack`, reinstalled `petsc` and `trilinos` loosing the support to `mumps` and here you go:
```
$ ctest -R "arpack"
Test project /Users/davydden/libs-sources/deal.ii/builds/parpack
    Start 1: arpack/step-36_ar.debug
1/6 Test #1: arpack/step-36_ar.debug ............................   Passed    1.28 sec
    Start 2: arpack/step-36_ar.release
2/6 Test #2: arpack/step-36_ar.release ..........................   Passed    0.39 sec
    Start 3: arpack/step-36_parpack.mpirun=3.debug
3/6 Test #3: arpack/step-36_parpack.mpirun=3.debug ..............   Passed    7.59 sec
    Start 4: arpack/step-36_parpack.mpirun=3.release
4/6 Test #4: arpack/step-36_parpack.mpirun=3.release ............   Passed    6.11 sec
    Start 5: arpack/step-36_parpack_trilinos.mpirun=3.debug
5/6 Test #5: arpack/step-36_parpack_trilinos.mpirun=3.debug .....   Passed    0.40 sec
    Start 6: arpack/step-36_parpack_trilinos.mpirun=3.release
6/6 Test #6: arpack/step-36_parpack_trilinos.mpirun=3.release ...   Passed    0.26 sec

100% tests passed, 0 tests failed out of 6

Total Test time (real) =  16.03 sec
```

Take away message - if there is no scalapack on the system, you are safe.

**Comment by tamiko** [comment]
*ugh* What a mess.

@davydden The only problem with that is that a fortran compile doesn't always have to be available.
With that we can either always fail if no fortran compiler is available, or break maybe half of all compilations that accidentally try to enable the parpack interface. An example is the 'tester' site that configures with Ninja (and thus has no fortran support available).

**Comment by davydden** [comment]
@tamiko you don't have to do fortran, just run one of the supplied unit test, e.g. `arpack/step-36_parpack_trilinos.mpirun=3.debug` and if it works - it is all good. As I wrote above, it's all good for me (tests pass) as long as there is no scalapack on the system.

**Comment by tamiko** [comment]
@davydden I cannot run a unit test in the build system. ;-) We'll discuss next break. :-]

**Comment by davydden** [comment]
of course, my bad ;-)))

**Comment by tamiko** [comment]
@davyden Please rebase and reindent :-)

All, we had a short discussion last week at the deal.II workshop about parpack and arpack. @davydden and me propose to review and merge this pull request [1] and - after that - redesign all interfaces of slepc/aprack/parpack to be uniform and avoid code duplication [2]. @oneliefleft Any update for #1117 ? :-)

[1] There is still a configure time check missing to detect a broken scalapack installation - I'll have a look at this.

[2] Which probably includes rewriting (p)arpack support to be provided by a single class.

**Comment by davydden** [comment]
btw, @tamiko you can run unit-test in this PR on a machine where `scalapack` is installed as a static library. In that case it shall not break `arpack`.

**Comment by tamiko** [comment]
@davydden Sorry for the long delay. I'll have a look at it in the next couple of days.

**Comment by davydden** [comment]
@tamiko any updates?

**Comment by davydden** [comment]
bwt, it turned out to be a parpack bug due to fortran sybmol collision: https://github.com/opencollab/arpack-ng/issues/18#issuecomment-135168310 and https://github.com/opencollab/arpack-ng/issues/18#issuecomment-136123261. Apparently someone copy-pasted an old version of that Scalapack function but did not change its name. However the newer version in the Scalapack 2.0.2 can't be used in the same way (they are two different functions now!). That led to SEGV.

**Comment by tamiko** [comment]
> @tamiko any updates?

I haven't had any time into looking into it :-/

Do you now in which version of (p)arpack it is resolved? We could just
very lazily link against parpack and scalapack, look for the symbol
clash - and if yes, configure without it.

**Comment by davydden** [comment]
> Do you now in which version of (p)arpack it is resolved? We could just
> very lazily link against parpack and scalapack, look for the symbol
> clash - and if yes, configure without it.

it's not resolved yet. If the PR i created in aprack will be merged, it will be resolved in the next version, which is `3.3.0`, AFAIK. That's the one which is currently maintained by Debian, Octave and Scilab https://github.com/opencollab/arpack-ng  as the original Fortran code is not maintained by anyone since end of 90s, AFAIK.

I agree with looking for a symbol clash (check for `pdlamch`). If `scalapack` is a dynamic library and the symbol clash is there, then `parpack` interface must be turned off. That does not have to happen if `scalapack` is static.

**Comment by davydden** [comment]
`arpack-ng` has merged the fix. @tamiko you can check that it all works now if you install arpack from HEAD. 

I would appreciate if we can finally merge it :-)

p.s. failing indentation checks are not mine, it was wrong in master when i rebased.

**Comment by tamiko** [comment]
@davydden Can you please apply the following?
```
diff --git a/include/deal.II/base/config.h.in b/include/deal.II/base/config.h.in
index 13684bb..9248f33 100644
--- a/include/deal.II/base/config.h.in
+++ b/include/deal.II/base/config.h.in
@@ -53,7 +53,7 @@
 #cmakedefine DEAL_II_WITH_TRILINOS
 #cmakedefine DEAL_II_WITH_UMFPACK
 #cmakedefine DEAL_II_WITH_ZLIB
-#cmakedefine DEAL_II_ARPACK_WITH_PARPACK
+
 
 /***********************************************************************
  * Compiler bugs:
@@ -135,11 +135,16 @@
 
 
 /***********************************************************************
- * Thread support
+ * Feature configuration
  *
- * For documentation see cmake/configure/configure_1_threads.cmake
+ * For documentation see cmake/configure/configure_*.cmake and
+ * cmake/modules/Find*.cmake
  */
 
+/* cmake/moduels/FindARPACK.cmake */
+#cmakedefine DEAL_II_ARPACK_WITH_PARPACK
+
+/* cmake/configure/configure_1_threads.cmake */
 #cmakedefine DEAL_II_USE_MT_POSIX
 #cmakedefine DEAL_II_USE_MT_POSIX_NO_BARRIERS
 
```

File: tests/arpack/step-36_parpack.cc [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+
+  std::vector<dealii::PETScWrappers::MPI::Vector> eigenfunctions;
+  std::vector<PetscScalar>                        eigenvalues;
+  dealii::PETScWrappers::MPI::SparseMatrix        stiffness_matrix, mass_matrix;
+
+  dealii::GridGenerator::hyper_cube (triangulation, -1, 1);
+  triangulation.refine_global (global_mesh_refinement_steps);
+
+  // METIS:
+  //dealii::GridTools::partition_triangulation (n_mpi_processes, triangulation);
```
Comment (tamiko) [inline tests/arpack/step-36_parpack.cc]: leftover :-)

File: tests/arpack/step-36_parpack_trilinos.cc [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+
+  std::vector<TrilinosWrappers::MPI::Vector> eigenfunctions;
+  std::vector<double>                                eigenvalues;
+  TrilinosWrappers::SparseMatrix             stiffness_matrix, mass_matrix;
+
+  GridGenerator::hyper_cube (triangulation, -1, 1);
+  triangulation.refine_global (global_mesh_refinement_steps);
+
+  //METIS:
+  //GridTools::partition_triangulation (n_mpi_processes, triangulation);
```
Comment (tamiko) [inline tests/arpack/step-36_parpack_trilinos.cc]: same here - or convert into comment describing that the following code is a manual partition and mimics above call.

**Comment by tamiko** [comment]
Can you verify that your parpack class has seen the same fix as was provided for the arpack solver by @guidokanschat? (Pull request #759)

**Comment by tamiko** [comment]
I suggest the following configure option:

```
diff --git a/cmake/configure/configure_arpack.cmake b/cmake/configure/configure_arpack.cmake
index 206b0cc..fa0e449 100644
--- a/cmake/configure/configure_arpack.cmake
+++ b/cmake/configure/configure_arpack.cmake
@@ -20,4 +20,4 @@
 SET(FEATURE_ARPACK_DEPENDS LAPACK)
 
 CONFIGURE_FEATURE(ARPACK)
-SET(DEAL_II_ARPACK_WITH_PARPACK ARPACK_WITH_PARPACK)
+SET(DEAL_II_ARPACK_WITH_PARPACK ${ARPACK_WITH_PARPACK})
diff --git a/cmake/modules/FindARPACK.cmake b/cmake/modules/FindARPACK.cmake
index b15a4c7..f8e26eb 100644
--- a/cmake/modules/FindARPACK.cmake
+++ b/cmake/modules/FindARPACK.cmake
@@ -52,6 +52,25 @@ ELSE()
   SET(PARPACK_LIBRARY "PARPACK_LIBRARY-NOTFOUND")
 ENDIF()
 
+IF(NOT DEAL_II_ARPACK_WITH_PARPACK)
+  #
+  # We have to avoid an unfortuante symbol clash with "libscalapack.so" -
+  # arpack happened to blindly copy a symbol name...
+  #   https://github.com/opencollab/arpack-ng/issues/18
+  #   https://github.com/opencollab/arpack-ng/pull/21
+  #
+  # Just disable parpack support if scalapack is present in Trilinos' or
+  # PETSc's link interface. This can be overridden by manually setting
+  # DEAL_II_ARPACK_WITH_PARPACK to true.
+  #
+  FOREACH(_libraries ${TRILINOS_LIBRARIES} ${PETSC_LIBRARIES})
+    IF("${_libraries}" MATCHES "scalapack")
+      SET(PARPACK_LIBRARY "PARPACK_LIBRARY-NOTFOUND")
+    ENDIF()
+  ENDFOREACH()
+ENDIF()
+
+
 IF(NOT PARPACK_LIBRARY MATCHES "-NOTFOUND")
   SET(ARPACK_WITH_PARPACK TRUE)
 ELSE()
```

**Comment by tamiko** [comment]
The reasoning behind this approach that anything else is infeasible :-/

I cannot check the full link interface because I refuse to restructure the build system for that :-)
Further, even if we can we would have to distinguish between
 - crash because of symbol clash
 - compilation error because no symbol is present (it got renamed...)

So, this is the lazy approach - we can always revisit this configure time check as soon as the new arpack version is widely used..

**Comment by davydden** [comment]
i think it's all good now. as to the fix, see this commit: https://github.com/davydden/dealii/commit/f0e08046dd06b428541447d9a9dece1be11873df

**Pull request merged**